imap_processing/mag/l1b/mag_l1b.py

```diff
@@ -0,0 +1,112 @@
+"""MAG L1B Processing."""
+
+from pathlib import Path
+
+import numpy as np
+import xarray as xr
+
+from imap_processing.cdf.imap_cdf_manager import ImapCdfAttributes
+from imap_processing.cdf.utils import load_cdf
+
+
+def mag_l1b(input_dataset: xr.Dataset, version: str) -> xr.Dataset:
+    """
+    Process MAG L1B data from L1A data.
+
+    Parameters
+    ----------
+    input_dataset : xr.Dataset
+        The input dataset to process
+    version : str
+        The version of the output data
+
+    Returns
+    -------
+    output_dataset : xr.Dataset
+        The processed dataset
+    """
+    # TODO:
+    # Read in calibration file
+    # multiply all vectors by calibration file
+
+    output_dataset = mag_l1b_processing(input_dataset)
+    attribute_manager = ImapCdfAttributes()
+    attribute_manager.add_instrument_global_attrs("mag")
+    attribute_manager.add_global_attribute("Data_version", version)
+
+    # Variable attributes can remain the same as L1A
+    input_logical_source = input_dataset.attrs["Logical_source"]
+    if isinstance(input_dataset.attrs["Logical_source"], list):
+        input_logical_source = input_dataset.attrs["Logical_source"][0]
+
+    logical_source = input_logical_source.replace("l1a", "l1b")
+    output_dataset.attrs = attribute_manager.get_global_attributes(logical_source)
+
+    return output_dataset
+
+
+def mag_l1b_processing(input_dataset: xr.Dataset) -> xr.Dataset:
+    """
+    Process MAG L1B data from L1A data.
+
+    MAG L1B is almost identical to L1A, with only the vectors and attributes getting
+    updated. All non-vector variables are the same.
+
+    Parameters
+    ----------
+    input_dataset : xr.Dataset
+        The input dataset to process
+    """
+    # TODO: There is a time alignment step that will add a lot of complexity.
+    # This needs to be done once we have some SPICE time data.
+
+    dims = [["direction"]]
+    new_dims = [["direction"]]
+    # TODO: This should definitely be loaded from AWS
+    calibration_dataset = load_cdf(
+        Path(__file__).parent / "imap_calibration_mag_20240229_v01.cdf"
+    )
+    # TODO: Check validity of time range for calibration
+    if "mago" in input_dataset.attrs["Logical_source"][0]:
+        calibration_matrix = calibration_dataset["MFOTOURFO"]
+    else:
+        calibration_matrix = calibration_dataset["MFITOURFI"]
+
+    l1b_fields = xr.apply_ufunc(
```

Comment:
You should be able to vectorize this as well I think rather than going through each vector individually. I still am not a fan of `apply_ufunc` ;) This will be way faster and it is a bit tricky to follow but with comments I think it can make quite a bit of sense for the future.

```python
ranges = input_vectors[:, 3].astype(int)  # N
calibration_matrices = calibration_dataset[:, :, ranges].T  # (N, 3, 3)
# grab the (N, 3) input vectors
# add a new axis to make it (N, 1, 3)
# matmul the outer two dimensions to produce (N, 1, 3) rotated vectors
# Remove the extra inner dimension
output_vectors = (input_vectors[:, :3][:, np.newaxis, :] @ calibration_matrices)[:, 0, :]
```

Reply:
There are future changes that I think would be useful to have apply_ufunc for, but I will keep this in mind for my next update if it ends up remaining this way. I agree it's a little overkill for what it is now.